### PR TITLE
Go: Better subPackages and excludedPackages

### PR DESCRIPTION
**Pull request by mmlb**
###### Motivation for this change

buildGo* functions had poor documentation for the `subPackages` and `excludedPackages` options, leading to incorrect usage. I've added documentation for them, updated the implementation of `excludedPackages` to match a common mis-use that actually seems easier to use. I've also updated the previously correct usage of `excludedPackages` (which is still correct) to the easier to manage list-of-strings that was previously incorrect.

I also got rid of instances of `testdata` in `excludedPackages` that were unnecessary.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I've built all the affected packages but not with nixpkgs-review (yet, its currently running) and verified functionality of a subset of the affected packages.

### CODE REVIEW

**Comment by zowoq** [comment]
Can you squash some of your commits please?

https://github.com/NixOS/nixpkgs/commit/d7d4df94c35d590c13b04e4e43b5ab15172d19a9 and https://github.com/NixOS/nixpkgs/commit/334afc0330b276872a911f8d6677040045e736ef should remain separate, everything else can be squashed into one commit.

File: pkgs/development/go-modules/generic/default.nix [hunk 1]
```diff
@@ -51,6 +51,12 @@
 , buildFlags ? ""
 , buildFlagsArray ? ""
 
+# string or list of strings used to filter out packages
+, excludedPackages ? ""
+
+# string or list of strings used to determine packages to build
+, subPackages ? ""
```
Comment (zowoq) [inline pkgs/development/go-modules/generic/default.nix:58]: Please explain why `excludedPackages` and `subPackages` need to be explicity included here (and in `go-packages`)?

Your commit message says:
> I had to read the implementation to figure out how to use subPackages which
should have been documented better to begin with. This would be better to have in
the manual, but this is at least better than before.

But doesn't give a technical reason for them to be here?

They shouldn't be added just for documentation and you have also added/updated them in the go docs so the reason stated in the commit message is no longer valid?

**Review by zowoq — CHANGES_REQUESTED** [review]

File: pkgs/development/go-modules/generic/default.nix [hunk 2]
```diff
@@ -171,13 +177,24 @@ let
...
       runHook preBuild
 
+      exclude='\(/_\|examples\|Godeps\|testdata'
+      if [[ -n "$excludedPackages" ]]; then
+        IFS=' ' read -r -a excludedArr <<<$excludedPackages
+        if ((''${#excludedArr[@]} > 1)); then
+          printf -v excludedAlternates '%s\\|' "''${excludedArr[@]}"
+          excludedAlternates=''${excludedAlternates%\\|} # drop final \| added by printf
+          exclude+='\\|'"$excludedAlternates"
+        else
```
Comment (zowoq) [inline pkgs/development/go-modules/generic/default.nix:187]: Does this need an if/else or can we do this?
```
      exclude='\(/_\|examples\|Godeps\|testdata'
      if [[ -n "$excludedPackages" ]]; then
        IFS=' ' read -r -a excludedArr <<<$excludedPackages
        printf -v excludedAlternates '%s\\|' "''${excludedArr[@]}"
        excludedAlternates=''${excludedAlternates%\\|} # drop final \| added by printf
        exclude+='\\|'"$excludedAlternates"
      fi
      exclude+='\)'
```

**Comment by mmlb** [comment]
> Can you squash some of your commits please?
> 
> [d7d4df9](https://github.com/NixOS/nixpkgs/commit/d7d4df94c35d590c13b04e4e43b5ab15172d19a9) and [334afc0](https://github.com/NixOS/nixpkgs/commit/334afc0330b276872a911f8d6677040045e736ef) should remain separate, everything else can be squashed into one commit.

I'll rework the commits and get rid of some of the intermediate stages. Do you mean for just 2 commits though? That seems like squashing down a bit to far to me tbh.

File: pkgs/development/go-modules/generic/default.nix (hunk 2 above)
Comment (mmlb) [reply pkgs/development/go-modules/generic/default.nix:187]: yes indeed, will update

File: pkgs/development/go-modules/generic/default.nix (hunk 1 above)
Comment (mmlb) [reply pkgs/development/go-modules/generic/default.nix:58]: Well its not just for documentation, these are inputs to the function and seems like they should show up explicitly instead of just coming in implicitly via the `...` right?

Comment (zowoq) [reply pkgs/development/go-modules/generic/default.nix:58]: No, please remove them.

**Comment by zowoq** [comment]
> I'll rework the commits and get rid of some of the intermediate stages. Do you mean for just 2 commits though? That seems like squashing down a bit to far to me tbh.

Three commits in total: https://github.com/NixOS/nixpkgs/commit/d7d4df94c35d590c13b04e4e43b5ab15172d19a9, https://github.com/NixOS/nixpkgs/commit/334afc0330b276872a911f8d6677040045e736ef can be left as they are. The the changes to `go-{modules,packages}` + docs can be squashed into one commit.

File: pkgs/development/go-modules/generic/default.nix (hunk 1 above)
Comment (mmlb) [reply pkgs/development/go-modules/generic/default.nix:58]: feels weird to me tbh, but ok.

**Comment by mmlb** [comment]
> > I'll rework the commits and get rid of some of the intermediate stages. Do you mean for just 2 commits though? That seems like squashing down a bit to far to me tbh.
> 
> Three commits in total. [d7d4df9](https://github.com/NixOS/nixpkgs/commit/d7d4df94c35d590c13b04e4e43b5ab15172d19a9), [334afc0](https://github.com/NixOS/nixpkgs/commit/334afc0330b276872a911f8d6677040045e736ef) can be left as they are. The changes to `go-{modules,packages}` + docs can be squashed into one commit.

also feels weird to me as I usually like to do one thing per commit, but ok.

File: pkgs/development/go-packages/generic/default.nix [hunk 3]
```diff
@@ -176,8 +193,8 @@ let
         if [ -n "$subPackages" ]; then
           echo "$subPackages" | sed "s,\(^\| \),\1$goPackagePath/,g"
         else
-          pushd "$NIX_BUILD_TOP/go/src" >/dev/null
-          find "$goPackagePath" -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort | uniq
+          pushd "$NIX_BUILD_TOP/go/src/$goPackagePath" >/dev/null
+          find . -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort --unique
```
Comment (zowoq) [inline pkgs/development/go-packages/generic/default.nix:197]: This is an unrelated change, please open a separate PR for this.

**Review by zowoq — CHANGES_REQUESTED** [review]

**Comment by mmlb** [comment]
Thanks for the feedback/reviews @zowoq I believe I've addressed everything in this last push.

**Comment by mmlb** [comment]
oops, forgot to get rid of the unnecessary `if`s. *Now* everything should be addressed.

**Comment by zowoq** [comment]
Using `excludedPackages` is much nicer now, thanks!

I'll leave this open for a couple of days for others to comment.

**Comment by zowoq** [comment]
Needs to go to staging due to the number of rebuilds, I'll watch for any conflicts or new uses of `excludedPackages` that need to be fixed in the meantime.

**Comment by yu-re-ka** [comment]
This breaks prometheus on staging-next

**Comment by zowoq** [comment]
I'm not really familiar with `prometheus`, does https://github.com/NixOS/nixpkgs/pull/166968#issuecomment-1086736602 work?

**Comment by mmlb** [comment]
https://github.com/NixOS/nixpkgs/pull/167116 fixes prometheus

**Comment by 06kellyjac** [comment]
I also ran into this with kubescape. I was trying to use excludedPackages and I was very confused when it wasn't working since I thought it had worked before

https://github.com/NixOS/nixpkgs/pull/168282/files#diff-1199902783e0ec4ac96a56a4e6219d055610efa991e6fc320da42d574ffb9dbeR35

---

is there any interest in moving `grep -q "$exclude" <<<$pkg && continue` into `buildGoDir` is this the way it'll be going forward?
